### PR TITLE
fix pad vs. eos token misidentification

### PR DESCRIPTION
**Pull request by aciddelgado**
Addresses this issue: #1682 regarding the misidentification of eos tokens as padding tokens. For models which share EOS and pad token, each EOS token present in the input would be identified as a pad token, leading to incorrect attention_mask and later parity issues. This PR fixes this issue by adding a special case for batch_size 1 (where input padding is unnecessary and assumed to not be there) as well as iterating through the input backwards in batch_size > 1 scenarios, as to not misinterpret any EOS tokens which are not at the end of the sequence.

### CODE REVIEW

File: test/c_api_tests.cpp [hunk 1]
```diff
@@ -447,9 +447,9 @@ TEST(CAPITests, EndToEndPhi) {
 
   // Verify outputs match expected outputs
   std::vector<int32_t> expected_output{
-      1212, 318, 257, 1332, 13, 198, 50280, 2, 16926, 1330, 1635, 10412, 6617, 278,
-      6335, 32994, 21857, 13849, 38665, 82, 21815, 1108, 9557, 40755, 27446, 2417,
-      6381, 6, 7131, 6, 14870, 31314, 21411, 46009, 3974, 82, 1039, 889, 263, 3684};
+      1212, 318, 257, 1332, 13, 50256, 198, 198, 198, 198, 4010, 4420, 43168, 15666,
```
Comment (sayanshaw24) [inline test/c_api_tests.cpp:489]: are these from HF? any chance you can add a comment above this with the model id from transformers and why the eos/padding changes affects the input ids in this way (for instance the changes to non-eos/padding ids, those should intuitively not change right) - so in the future we know why this is expected.

File: test/c_api_tests.cpp [hunk 2]
```diff
@@ -427,7 +427,7 @@ TEST(CAPITests, EndToEndPhi) {
   auto model = OgaModel::Create(PHI2_PATH);
   auto tokenizer = OgaTokenizer::Create(*model);
 
-  const char* input_string = "This is a test.";
+  const char* input_string = "This is a test.<|endoftext|>";
```
Comment (kunal-vaishnavi) [inline test/c_api_tests.cpp:469]: Is this to get the EOS token id in the tokenized input?

Comment (aciddelgado) [reply test/c_api_tests.cpp:469]: Yes

File: test/c_api_tests.cpp (hunk 1 above)
Comment (aciddelgado) [reply test/c_api_tests.cpp:489]: I had some trouble understanding your comment but basically this should be appending an EOS token at the end which should a) change the output (at a rudimentary level the model can interpret this as a change of turn in conversation) and b) this eos token would be masked to 0 rather than 1 without this pr, which is a mistake as the EOS token is valid and should not be masked out.

Comment (sayanshaw24) [reply test/c_api_tests.cpp:489]: awesome, thanks! by HF i just meant where are we getting the expected input ids from (for instance from HuggingFace transformers, model id ___) could you just add a quick comment above the specific test along with this context you gave so we know what changed in the future if/when editing this file?
